File: app/src/main/java/es/etologic/mahjongscoring2/app/old_games/OldGamesFragment.java
package es.etologic.mahjongscoring2.app.old_games;

import android.app.AlertDialog;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import es.etologic.mahjongscoring2.Injector;
import es.etologic.mahjongscoring2.R;
import es.etologic.mahjongscoring2.app.main.MainActivityViewModel;
import es.etologic.mahjongscoring2.app.model.ShowState;
import es.etologic.mahjongscoring2.domain.entities.Game;

import static android.view.View.VISIBLE;
import static es.etologic.mahjongscoring2.app.model.MainScreens.NEW_GAME;
import static es.etologic.mahjongscoring2.app.model.ShowState.HIDE;
import static es.etologic.mahjongscoring2.app.model.ShowState.SHOW;

public class OldGamesFragment extends Fragment implements OldGamesRvAdapter.GameItemListener {

    //VIEWS
    @BindView (R.id.toolbarOldGames) Toolbar toolbar;
    @BindView (R.id.swipeRefreshLayoutOldGames) SwipeRefreshLayout swipeRefreshLayout;
    @BindView (R.id.recyclerViewOldGames) RecyclerView recyclerView;
    @BindView (R.id.emptyLayoutOldGames) LinearLayout emptyLayout;
    //FIELDS
    private Unbinder unbinder;
    private OldGamesRvAdapter rvAdapter;
    private Context context;
    private MainActivityViewModel activityViewModel;
    private OldGamesViewModel viewModel;

    //LIFECYCLE
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.oldgames_fragment, container, false);
    }
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        context = getContext();
        unbinder = ButterKnife.bind(this, view);
        setupRecyclerView();
        setupViewModel();
        setupSwipeRefreshLayout();
        setToolbar();
    }
    private void setupViewModel() {
        activityViewModel = ViewModelProviders
                .of(getActivity(), Injector.provideMainActivityViewModelFactory())
                .get(MainActivityViewModel.class);
        viewModel = ViewModelProviders
                .of(this, Injector.provideOldGamesViewModelFactory(context))
                .get(OldGamesViewModel.class);
        viewModel.getOldGames().observe(this, this :: setGames);
        viewModel.getProgressState().observe(this, this :: toogleLocalProgress);
        viewModel.getSnackbarMessage().observe(this, this :: showSnackbar);
    }
    private void setupSwipeRefreshLayout() {
        swipeRefreshLayout.setColorSchemeColors(getResources().getIntArray(R.array.swipeRefreshColors));
        swipeRefreshLayout.setOnRefreshListener(() -> viewModel.loadOldGames());
    }
    private void setupRecyclerView() {
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        rvAdapter = new OldGamesRvAdapter();
        rvAdapter.setOldGameItemListener(this);
        recyclerView.setAdapter(rvAdapter);
    }
    private void setToolbar() { activityViewModel.setToolbar(toolbar); }
    private void setGames(List<Game> games) {
        if(games == null || games.isEmpty()) {
            emptyLayout.setVisibility(VISIBLE);
        } else {
            if(emptyLayout.getVisibility() == VISIBLE) emptyLayout.setVisibility(View.GONE);
            rvAdapter.setGames(games);
            toogleLocalProgress(HIDE);
        }
    }
    private void toogleLocalProgress(ShowState showState) { swipeRefreshLayout.setRefreshing(showState == SHOW); }
    private void showSnackbar(String message) { Snackbar.make(toolbar, message, Snackbar.LENGTH_LONG).show(); }
    @Override
    public void onResume() {
        super.onResume();
        viewModel.loadOldGames();
    }
    @Override
    public void onDestroy() {
        unbinder.unbind();
        super.onDestroy();
    }

    //EVENTS
    @OnClick (R.id.fabOldGames) public void onFabNewGameClick() { activityViewModel.goToScreen(NEW_GAME); }
    @Override
    public void onOldGameItemDeleteClicked(long gameId) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle(R.string.delete_game)
                .setMessage(R.string.are_you_sure)
                .setPositiveButton(R.string.delete, (dialog, which) -> viewModel.deleteGame(gameId))
                .setNegativeButton(android.R.string.cancel, null)
                .create()
                .show();
    }
    @Override
    public void onOldGameItemResumeClicked(long gameId) { activityViewModel.goToGame(gameId); }
}

File: app/src/main/java/es/etologic/mahjongscoring2/app/main/old_games/OldGamesFragment.kt
package es.etologic.mahjongscoring2.app.main.old_games

import android.app.AlertDialog
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView.*
import com.google.android.material.snackbar.Snackbar
import es.etologic.mahjongscoring2.R
import es.etologic.mahjongscoring2.app.base.BaseFragment
import es.etologic.mahjongscoring2.app.main.activity.MainActivityViewModel
import es.etologic.mahjongscoring2.app.main.activity.MainActivityViewModelFactory
import es.etologic.mahjongscoring2.app.model.ShowState
import es.etologic.mahjongscoring2.app.model.ShowState.HIDE
import es.etologic.mahjongscoring2.app.model.ShowState.SHOW
import es.etologic.mahjongscoring2.domain.model.GameWithRounds
import kotlinx.android.synthetic.main.main_oldgames_fragment.*
import javax.inject.Inject

class OldGamesFragment : BaseFragment(), OldGamesRvAdapter.GameItemListener {
    
    @Inject
    internal lateinit var rvAdapter: OldGamesRvAdapter
    @Inject
    internal lateinit var mainActivityViewModelFactory: MainActivityViewModelFactory
    @Inject
    internal lateinit var oldGamesViewModelFactory: OldGamesViewModelFactory
    private lateinit var activityViewModel: MainActivityViewModel
    private lateinit var viewModel: OldGamesViewModel
    
    //EVENTS
    override fun onOldGameItemDeleteClicked(gameId: Long) {
        val builder = AlertDialog.Builder(activity)
        builder.setTitle(R.string.delete_game)
            .setMessage(R.string.are_you_sure)
            .setPositiveButton(R.string.delete) { _, _ -> viewModel.deleteGame(gameId) }
            .setNegativeButton(android.R.string.cancel, null)
            .create()
            .show()
    }
    
    override fun onOldGameItemResumeClicked(gameId: Long) {
        activityViewModel.startGame(gameId)
    }
    
    //LIFECYCLE
    override fun onResume() {
        super.onResume()
        viewModel.getAllGames()
    }
    
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.main_oldgames_fragment, container, false)
    }
    
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupRecyclerView()
        setupViewModel()
        setupSwipeRefreshLayout()
        setToolbar()
        setOnClickListeners()
    }
    
    private fun setupRecyclerView() {
        recyclerViewOldGames?.setHasFixedSize(true)
        val layoutManager = LinearLayoutManager(context)
        recyclerViewOldGames?.layoutManager = layoutManager
        rvAdapter.setOldGameItemListener(this)
        recyclerViewOldGames?.adapter = rvAdapter
    }
    
    private fun setupViewModel() {
        if (activity != null) {
            activityViewModel = ViewModelProvider(activity!!, mainActivityViewModelFactory).get(MainActivityViewModel::class.java)
            viewModel = ViewModelProvider(this, oldGamesViewModelFactory).get(OldGamesViewModel::class.java)
            viewModel.getError().observe(viewLifecycleOwner, Observer(this::showError))
            viewModel.getGames().observe(viewLifecycleOwner, Observer(this::setGames))
            viewModel.getSnackbarMessage().observe(viewLifecycleOwner, Observer(this::showSnackbar))
            viewModel.getProgressState().observe(viewLifecycleOwner, Observer(this::toogleLocalProgress))
        }
    }
    
    private fun setGames(games: List<GameWithRounds>) {
        if (games.isEmpty())
            emptyLayoutOldGames?.visibility = VISIBLE
        else {
            emptyLayoutOldGames?.visibility = GONE
            rvAdapter.setGames(games)
        }
        toogleLocalProgress(HIDE)
    }
    
    private fun toogleLocalProgress(showState: ShowState) {
        swipeRefreshLayoutOldGames?.isRefreshing = showState === SHOW
    }
    
    private fun showSnackbar(message: String) {
        Snackbar.make(toolbarOldGames, message, Snackbar.LENGTH_LONG).show()
    }
    
    private fun setupSwipeRefreshLayout() {
        if (activity != null) {
            swipeRefreshLayoutOldGames?.setColorSchemeColors(
                ContextCompat.getColor(activity!!, R.color.colorPrimary),
                ContextCompat.getColor(activity!!, R.color.colorAccent),
                ContextCompat.getColor(activity!!, R.color.purplePenalty),
                ContextCompat.getColor(activity!!, R.color.colorPrimaryDark)
            )
            swipeRefreshLayoutOldGames?.setOnRefreshListener { viewModel.getAllGames() }
        }
    }
    
    private fun setToolbar() {
        activityViewModel.setToolbar(toolbarOldGames)
    }
    
    private fun setOnClickListeners() {
        fabOldGames?.setOnClickListener {
            activityViewModel.startNewGame()
        }
    }
}
